src/field/FieldDriver.i.hh

```diff
@@ -0,0 +1,269 @@
+                         RungeKuttaStepper<MagFieldEquation>& stepper)
+    : shared_(shared), stepper_(stepper)
+{
+    CELER_ENSURE(shared_);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Adaptive step control based on G4ChordFinder and G4MagIntegratorDriver:
+ * For a given trial step, advance by a sub-step within a required tolerance
+ * and update the current state (position and momentum).  For an efficient
+ * adaptive integration, the proposed chord of which the miss-distance (the
+ * closest distance from the curved trajectory to the chord) is smaller than
+ * a reference distance (dist_chord) will be accepted if its stepping error is
+ * within a reference accuracy. Otherwise, the more accurate step integration
+ * (advance_accurate) will be performed.
+ */
+CELER_FUNCTION
+real_type FieldDriver::operator()(real_type step, OdeState* state)
+{
+    // Output with a step control error
+    FieldOutput output = this->find_next_chord(step, *state);
+
+    real_type step_taken = output.step_taken;
+
+    // Evaluate the relative error
+    real_type rel_error = output.value / (shared_.epsilon_step * step_taken);
+
+    if (rel_error > 1)
+    {
+        // Advance more accurately with a newly proposed step
+        real_type next_step = this->new_step_size(step, rel_error);
+        step_taken
+            = this->accurate_advance(step_taken, &output.state, next_step);
+    }
+
+    // Accept this accuracy and update the current state
+    *state = output.state;
+
+    return step_taken;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Find the next acceptable chord of which the miss-distance is smaller than
+ * a given reference (delta_chord) and evaluate the associated error.
+ */
+CELER_FUNCTION auto
+FieldDriver::find_next_chord(real_type step, const OdeState& state)
+    -> FieldOutput
+{
+    // Output with a step control error
+    FieldOutput output;
+
+    // Try with the proposed step
+    output.step_taken = step;
+
+    bool          converged       = false;
+    unsigned int  remaining_steps = shared_.max_nsteps;
+    StepperResult result;
+
+    do
+    {
+        result = stepper_(step, state);
+
+        // Check whether the distance to the chord is small than the reference
+        real_type dchord
+            = distance_chord(state, result.mid_state, result.end_state);
+
+        if (dchord <= shared_.delta_chord + rel_tolerance())
+        {
+            converged    = true;
+            output.value = truncation_error(
+                step, shared_.epsilon_rel_max, state, result.err_state);
+        }
+        else
+        {
+            // Estimate a new trial chord with a relative scale
+            output.step_taken
+                *= std::fmax(std::sqrt(shared_.delta_chord / dchord), half());
+        }
+    } while (!converged && (--remaining_steps > 0));
+
+    CELER_ASSERT(converged);
```

Comment:
Is that the intent that we trigger this assert if `remainging_steps` get to 0?

Reply:
If it stays that `converged` and `remaining_steps` are only used in the condition, an alternative could be 
```
do
{
       ....
        --remaining_steps;
        if (dchord <= (shared_.delta_chord + rel_tolerance()))
        {
              remaining_steps = 0;
              ...
         }
         else
         ....
} while (remaining_steps > 0);
```

Reply:
Actually the assert is checked to detect any potential rare case (with a non-uniform magnetic field and the more complicated geometry) that the loop is not converged.  I added "TODO with a comment" for now.